File: public/js/discover.js
/* =========================================================
   discover.js — 소재 탐색 (POST /api/discover)
   목록에 없는 소재를 입력하면 위키백과 + 웹 검색으로 자료를 모아
   활동 모듈 초안을 만들어 옵니다. 저장하면 '내 모듈'이 됩니다.
   ========================================================= */

const MyModules = {
  key: "mujacheonseo.modules.v1",

  all() {
    try {
      const raw = localStorage.getItem(this.key);
      return raw ? JSON.parse(raw) : [];
    } catch (e) { return []; }
  },

  save(mod) {
    const list = this.all();
    if (!mod.id) mod.id = "MY" + (list.length + 1) + "_" + Date.now().toString(36).slice(-4);
    mod.custom = true;
    list.unshift(mod);
    try { localStorage.setItem(this.key, JSON.stringify(list)); }
    catch (e) { toast("저장 공간이 부족합니다."); return null; }
    return mod;
  },

  remove(id) {
    localStorage.setItem(this.key, JSON.stringify(this.all().filter(function (m) { return m.id !== id; })));
  }
};

/* 기본 모듈 + 내 모듈 */
function allModules() {
  return MODULES.concat(MyModules.all());
}

let lastDiscovered = null;
let lastSources = [];

function initDiscover() {
  const sel = document.getElementById("d-target");
  sel.innerHTML = '<option value="">대상 (선택)</option>' +
    AGE_GROUPS.map(function (a) { return "<option>" + a.name + "</option>"; }).join("");

  document.getElementById("discoverForm").addEventListener("submit", onDiscover);
}

async function onDiscover(e) {
  e.preventDefault();
  const input = document.getElementById("d-keyword");
  const keyword = input.value.trim();

  input.classList.remove("is-invalid");
  if (!keyword) {
    input.classList.add("is-invalid");
    input.focus();
    alert("찾아볼 소재를 입력해 주세요!");
    return;
  }

  const btn = document.getElementById("btnDiscover");
  btn.disabled = true;
  document.getElementById("discoverLoading").hidden = false;
  document.getElementById("discoverError").hidden = true;
  document.getElementById("discoverResult").hidden = true;

  const controller = new AbortController();
  const timer = setTimeout(function () { controller.abort(); }, APP.apiTimeoutMs);
  const slow = setTimeout(function () {
    document.getElementById("discoverLoadingText").textContent =
      "응답이 지연되고 있습니다. 잠시만 기다려 주세요...";
  }, 12000);

  try {
    const res = await fetch("/api/discover", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        keyword: keyword,
        useSearch: document.getElementById("d-search").checked,
        target: document.getElementById("d-target").value,
        season: document.getElementById("f-season") ? document.getElementById("f-season").value : "",
        themes: Array.from(document.querySelectorAll("#themePicker .theme-chip.is-on"))
          .map(function (c) { const t = findTheme(c.dataset.theme); return t ? t.name : ""; })
      }),
      signal: controller.signal
    });

    if (!res.ok) {
      let detail = "";
      try { detail = (await res.json()).message || ""; } catch (_) {}
      return showDiscoverError(detail ||
        "AI 서버와의 통신이 원활하지 않습니다. 잠시 후 다시 시도해 주세요.");
    }

    const data = await res.json();
    if (!data || !data.module) return showDiscoverError("결과를 읽지 못했습니다. 다시 시도해 주세요.");

    lastDiscovered = data.module;
    lastSources = data.sources || [];
    renderDiscovered(lastDiscovered, lastSources);

  } catch (err) {
    if (err.name === "AbortError") {
      showDiscoverError("45초 안에 답이 오지 않아 요청을 중단했습니다. 다시 시도해 주세요.");
    } else {
      showDiscoverError("AI 서버와의 통신이 원활하지 않습니다. (" + err.message + ")");
    }
  } finally {
    clearTimeout(timer);
    clearTimeout(slow);
    btn.disabled = false;
    document.getElementById("discoverLoading").hidden = true;
    document.getElementById("discoverLoadingText").textContent = "자료를 모으는 중입니다…";
  }
}

function showDiscoverError(msg) {
  document.getElementById("discoverErrorMsg").textContent = msg;
  document.getElementById("discoverError").hidden = false;
}

function renderDiscovered(m, sources) {
  const el = document.getElementById("discoverResult");
  el.innerHTML =
    '<div class="card">' +
      '<span class="module-tag">새 모듈 초안</span>' +
      "<h3>" + esc(m.activity || m.name || "") + "</h3>" +
      '<p class="module-meta"><b>생태 소재</b> · ' + esc(m.subject || "") + "</p>" +
      '<p class="module-meta"><b>철학 배경</b> · ' + esc(m.philosophy || "") + "</p>" +
      '<p class="quote" style="font-size:.95rem;margin:.7rem 0">“' + esc(m.question || "") + "”</p>" +
      "<ul>" + (m.steps || []).map(function (s) { return "<li>" + esc(s) + "</li>"; }).join("") + "</ul>" +
      '<p class="module-meta" style="margin-top:.7rem"><b>준비물</b> · ' + esc((m.materials || []).join(", ")) + "</p>" +
      '<p class="module-meta"><b>안전</b> · ' + esc(m.safety || "") + "</p>" +
      (m.caution ? '<p class="caution-box">⚠ ' + esc(m.caution) + "</p>" : "") +
      (m.factNote ? '<p class="hint">📌 ' + esc(m.factNote) + "</p>" : "") +
      sourcesHTML(sources) +
      '<div class="result-toolbar" style="margin-top:1rem;border:0;padding:0">' +
        '<button class="btn btn-primary btn-sm" id="btnSaveModule">＋ 내 모듈로 저장</button>' +
        '<button class="btn btn-sm" id="btnUseModule">수업 설계에 바로 쓰기</button>' +
      "</div>" +
    "</div>";
  el.hidden = false;

  document.getElementById("btnSaveModule").addEventListener("click", function () {
    const saved = MyModules.save(Object.assign({}, lastDiscovered, { sources: lastSources }));
    if (!saved) return;
    renderLibrary();
    refreshModuleSelect();
    toast("내 모듈로 저장했습니다.");
  });

  document.getElementById("btnUseModule").addEventListener("click", function () {
    const saved = MyModules.save(Object.assign({}, lastDiscovered, { sources: lastSources }));
    if (!saved) return;
    renderLibrary();
    refreshModuleSelect();
    document.getElementById("f-module").value = saved.id;
    document.getElementById("f-module").dispatchEvent(new Event("change"));
    navigate("design");
    toast("설계 폼에 실었습니다.");
  });
}

/* 출처 목록 — 무엇을 근거로 만들었는지 보여 줍니다 */
function sourcesHTML(sources) {
  if (!sources || !sources.length) return "";
  return '<p class="module-meta" style="margin-top:.8rem"><b>참고한 자료</b></p><ul class="source-list">' +
    sources.map(function (s) {
      return "<li>" + (s.url
        ? '<a href="' + esc(s.url) + '" target="_blank" rel="noopener">' + esc(s.title) + "</a>"
        : esc(s.title)) + "</li>";
    }).join("") + "</ul>";
}

File: api/discover.py
# =========================================================
# api/discover.py — 소재 탐색 (POST /api/discover)
#
# 앱에 없는 소재를 사용자가 직접 입력하면,
#   ① 위키백과에서 배경 자료를 모으고
#   ② 구글 검색 그라운딩으로 최신 근거를 찾아
#   ③ 앱의 활동 모듈 형식으로 초안을 만들어 돌려줍니다.
#
# 정해진 20종 목록에 갇히지 않고 어떤 소재로도 수업을 설계하기 위한 엔드포인트입니다.
# =========================================================

import json
import os
import sys as _sys

_sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))  # 같은 폴더의 _common.py 를 찾기 위해

from _common import call_gemini, fetch_wikipedia

SYSTEM_RULES = """당신은 20년 경력의 숲해설 전문강사이자 생태 인문학 교육 설계자입니다.
사용자가 건넨 자연 소재 하나를 '활동 모듈' 한 장으로 정리합니다.

지켜야 할 원칙
1. 지식 전달형 설명이 아니라, 참가자가 직접 만지고 세어 보고 몸으로 흉내 내는 활동으로 짠다.
2. 생태적 사실을 먼저 확인하고, 거기서 자연스럽게 피어나는 인문학적 질문을 붙인다.
   억지로 고전을 끌어다 붙이지 않는다.
3. 안전 지침은 그 활동에서 실제로 일어날 수 있는 사고(독성, 알레르기, 미끄러짐, 쏘임,
   접근 금지 구역)를 기준으로 구체적으로 쓴다. "조심한다" 같은 막연한 문장은 쓰지 않는다.
4. 종 동정이나 사실 관계가 불확실하면 caution 에 무엇을 확인해야 하는지 적는다.
5. 출처를 확인할 수 없는 구체 수치(흡수량, 개체 수, 면적, 연도)는 쓰지 않는다.
6. 보호수·천연기념물·사유지일 가능성이 있으면 caution 에 사전 확인을 적는다.
7. 모든 문장은 한국어 존댓말, 현장 해설가가 바로 쓸 수 있는 실무 문장으로 쓴다.

반드시 아래 JSON 하나만 출력합니다. 설명이나 코드펜스를 덧붙이지 않습니다.

{
  "name": "모듈 주제명 (짧게)",
  "subject": "생태 소재 (종명 또는 대상)",
  "philosophy": "연결되는 철학·인문 배경",
  "question": "아이들에게 던질 대표 화두 한 문장",
  "activity": "활동명",
  "steps": ["활동 순서 4단계"],
  "types": ["오감 관찰|신체 모사·놀이|자연물 예술|글쓰기·시|이야기 나눔|측정·기록 중 해당하는 것"],
  "themes": ["art|culture|philosophy|music|science|engineering|history|literature|healing 중 해당하는 것"],
  "season": ["봄|여름|가을|겨울 중 관찰 가능한 계절"],
  "materials": ["준비물"],
  "safety": "구체적인 안전 지침 2~3문장",
  "target": "권장 대상",
  "caution": "해설 전 확인해야 할 사실 관계. 없으면 빈 문자열",
  "factNote": "이 모듈에서 사실로 확인한 내용과 확인하지 못한 내용을 각각 한 줄로"
}"""


def build_prompt(data, wiki):
    lines = [
        "다음 소재로 숲해설 활동 모듈 초안을 만들어 주세요.",
        "",
        "- 소재: %s" % data.get("keyword", ""),
    ]
    if data.get("target"):
        lines.append("- 학습 대상: %s" % data["target"])
    if data.get("season"):
        lines.append("- 계절: %s" % data["season"])
    if data.get("themes"):
        lines.append("- 관심사: %s" % ", ".join(data["themes"]))
    if data.get("note"):
        lines.append("- 해설가의 추가 요청: %s" % data["note"])

    if wiki:
        lines += ["", "참고 자료 (위키백과 발췌 — 사실 확인용으로만 쓰고 그대로 옮기지 마세요):"]
        for w in wiki:
            lines.append("  [%s] %s" % (w["title"], w["extract"]))

    lines += [
        "",
        "이 소재가 우리나라 숲에서 실제로 관찰 가능한지 먼저 확인하고,",
        "관찰이 어려운 소재라면 대체 관찰 대상을 caution 에 적어 주세요.",
    ]
    return "\n".join(lines)


# ---------------------------------------------------------
# 처리 함수 — api/index.py 가 호출합니다
# ---------------------------------------------------------
def handle(data):
    keyword = (data.get("keyword") or "").strip()
    if not keyword:
        return 400, {"ok": False, "message": "찾아볼 소재를 입력해 주세요!"}
    if len(keyword) > 60:
        return 400, {"ok": False, "message": "소재 이름이 너무 깁니다. (60자 이내)"}

    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        return 401, {"ok": False,
                     "message": "서버에 GEMINI_API_KEY 환경 변수가 설정되어 있지 않습니다."}

    # ① 위키백과에서 배경 자료 수집
    wiki = fetch_wikipedia(keyword)

    # ② 구글 검색 그라운딩 + 생성
    module, sources, err = call_gemini(api_key, SYSTEM_RULES, build_prompt(data, wiki))
    if err:
        return err[0], {"ok": False, "message": err[1]}

    wiki_sources = [{"title": w["title"] + " (위키백과)", "url": w["url"]} for w in wiki if w["url"]]

    module.setdefault("steps", [])
    module.setdefault("types", ["오감 관찰"])
    module.setdefault("themes", ["science"])
    module.setdefault("season", ["봄", "여름", "가을"])
    module.setdefault("materials", [])
    module.setdefault("caution", "")
    module["custom"] = True

    return 200, {"ok": True,
                 "model": os.environ.get("GEMINI_MODEL", "gemini-2.5-flash"),
                 "module": module,
                 "sources": wiki_sources + sources}
